File: .spec/scripts/python/detect_remote.py
#!/usr/bin/env python3
"""
detect_remote.py — Detect Git remote URL and parse GitHub owner/repo
(analogous to detect-remote.ps1).

Usage:
    python detect_remote.py [--json]

Exit code 0 always (graceful degradation when no remote).
"""

from __future__ import annotations

import argparse
import json
import re
import subprocess
import sys
from pathlib import Path


def _no_remote(reason: str, use_json: bool) -> None:
    """Emit a no-remote response and exit successfully."""
    if use_json:
        print(json.dumps({"has_remote": False, "is_github": False, "reason": reason}))
    else:
        print(f"[specify] {reason}", file=sys.stderr)
    sys.exit(0)


def main() -> None:
    """Detect remote.origin URL and parse GitHub owner/repo when possible."""
    parser = argparse.ArgumentParser(
        description="Detect Git remote URL and parse GitHub owner/repo."
    )
    parser.add_argument("--json", action="store_true", help="Output in JSON format")
    args = parser.parse_args()
    use_json: bool = args.json

    # Check git availability
    try:
        subprocess.run(["git", "--version"], capture_output=True, check=False)
    except FileNotFoundError:
        _no_remote("Git not found; cannot determine remote URL", use_json)

    # Check inside a git repo
    result = subprocess.run(
        ["git", "rev-parse", "--is-inside-work-tree"],
        capture_output=True, text=True
    )
    if result.returncode != 0:
        _no_remote("Not inside a Git repository; cannot determine remote URL", use_json)

    # Get remote URL
    result = subprocess.run(
        ["git", "config", "--get", "remote.origin.url"],
        capture_output=True, text=True
    )
    remote_url = result.stdout.strip()
    if not remote_url:
        _no_remote("No remote.origin configured", use_json)

    is_github = False
    owner = ""
    repo  = ""

    # HTTPS: https://github.com/<owner>/<repo>[.git]
    https_match = re.match(r'^https://github\.com/([^/]+)/([^/]+?)(?:\.git)?$', remote_url)
    if https_match:
        is_github, owner, repo = True, https_match.group(1), https_match.group(2)
    else:
        # SSH: git@github.com:<owner>/<repo>[.git]
        ssh_match = re.match(r'^git@github\.com:([^/]+)/([^/]+?)(?:\.git)?$', remote_url)
        if ssh_match:
            is_github, owner, repo = True, ssh_match.group(1), ssh_match.group(2)

    if use_json:
        print(json.dumps({
            "has_remote": True,
            "is_github":  is_github,
            "remote_url": remote_url,
            "owner":      owner,
            "repo":       repo,
        }, ensure_ascii=False))
    else:
        print(f"Remote URL: {remote_url}")
        if is_github:
            print(f"GitHub repository: {owner}/{repo}")


if __name__ == "__main__":
    main()

File: .spec/scripts/detect_remote.ts
#!/usr/bin/env node
import { spawnSync } from "node:child_process";

/**
 * Emits remote detection diagnostics to stderr.
 */
function logInfo(message: string, details?: unknown): void {
  if (details === undefined) {
    console.error(`[detect-remote] ${message}`);
    return;
  }
  console.error(`[detect-remote] ${message}`, details);
}

/**
 * Terminates with a non-error "no remote" outcome.
 */
function exitWithoutRemote(reason: string, useJson: boolean): never {
  logInfo("Remote detection ended without a remote", { reason, useJson });
  if (useJson) {
    console.log(JSON.stringify({ has_remote: false, is_github: false, reason }));
  } else {
    console.error(`[specify] ${reason}`);
  }
  process.exit(0);
}

/**
 * Parses known GitHub HTTPS/SSH origin URL formats.
 */
function parseGithubRemote(remoteUrl: string): { isGithub: boolean; owner: string; repo: string } {
  const httpsMatch = remoteUrl.match(/^https:\/\/github\.com\/([^/]+)\/([^/]+?)(?:\.git)?$/);
  if (httpsMatch) {
    return { isGithub: true, owner: httpsMatch[1], repo: httpsMatch[2] };
  }

  const sshMatch = remoteUrl.match(/^git@github\.com:([^/]+)\/([^/]+?)(?:\.git)?$/);
  if (sshMatch) {
    return { isGithub: true, owner: sshMatch[1], repo: sshMatch[2] };
  }

  return { isGithub: false, owner: "", repo: "" };
}

const useJson = process.argv.includes("--json");
logInfo("Starting remote detection", { useJson });

if (spawnSync("git", ["--version"], { stdio: "ignore" }).error) {
  exitWithoutRemote("Git not found; cannot determine remote URL", useJson);
}

const inWorkTreeResult = spawnSync("git", ["rev-parse", "--is-inside-work-tree"], { encoding: "utf-8" });
if (inWorkTreeResult.status !== 0) exitWithoutRemote("Not inside a Git repository; cannot determine remote URL", useJson);

const remoteUrlResult = spawnSync("git", ["config", "--get", "remote.origin.url"], { encoding: "utf-8" });
const remoteUrl = (remoteUrlResult.stdout ?? "").trim();
if (!remoteUrl) exitWithoutRemote("No remote.origin configured", useJson);
logInfo("Found remote origin URL", { remoteUrl });

const { isGithub, owner, repo } = parseGithubRemote(remoteUrl);
logInfo("Parsed remote metadata", { isGithub, owner: owner || null, repo: repo || null });

if (useJson) {
  console.log(JSON.stringify({ has_remote: true, is_github: isGithub, remote_url: remoteUrl, owner, repo }));
} else {
  console.log(`Remote URL: ${remoteUrl}`);
  if (isGithub) console.log(`GitHub repository: ${owner}/${repo}`);
}

